docs/source/core_concepts/index.rst

```diff
@@ -0,0 +1,512 @@
+  and state 2 is the final state.
+- There are 2 arcs. The first arc is from state 0 to state 1
+  with label 10 and score 0.1. The second arc is from state 1 to
+  the final state 2 with label -1 and score 0.2.
+
+.. HINT::
+
+  We use arc weight and arc score interchangeably in k2.
+
+The above FSA is created with the following code:
+
+.. code-block:: python
+
+    import k2
+    s = '''
+    0 1 10 0.1
+    1 2 -1 0.2
+    2
+    '''
+    fsa = k2.Fsa.from_str(s)
+    fsa.draw('simple_fsa.svg')
+
+We summarize the **unique** features of FSA in k2 in below:
+
+  - There is only one start state
+  - The start state is **always** 0
+  - All other states have a state number greater than 0
+  - There is only one final state
+  - The final state **always** has the **largest** state number
+  - Arcs entering the final state **always** have -1 as the label
+  - Arcs that do not enter the final state cannot have -1 as the label
+
+.. HINT::
+
+  Different from other frameworks, FSAs in k2 have only a single final state.
+
+  If you want to convert an FSA from another framework to k2 that contains
+  multiple final states, you can create an extra state and consider it as
+  the super final state. For each final state in the FSA, add an arc to this
+  super final state with label -1 and score 0. The resulting FSA will contain
+  only a single final state.
+
+  Similarly, if it contains multiple start states, you can add a super start
+  state and set both the label and score of the arcs added from the super start
+  state to the start state to 0.
+
+Attributes
+----------
+
+Arbitrary attributes can be attached to the arcs of an FSA.
+For example, we can attach a tensor attribute to an FSA indicating
+the output label of arcs so that the FSA is converted to an FST.
+
+The attached attributes are **automaticaly propagated** through operations,
+with ``autograd`` if they are real-valued tensors.
+
+The following code converts the above simple acceptor to a transducer:
+
+.. code-block:: python
+
+    import k2
+    import torch
+    s = '''
+    0 1 10 0.1
+    1 2 -1 0.2
+    2
+    '''
+    fsa = k2.Fsa.from_str(s)
+    fsa.aux_labels = torch.tensor([100, -1], dtype=torch.int32)
+    fsa.draw('simple_fst.svg')
+
+
+The resulting FST is visualized in :numref:`a simple fst`.
+
+.. _a simple fst:
+.. figure:: images/simple_fst.svg
+    :alt: A simple FST in k2
+    :align: center
+    :figwidth: 600px
+
+    A simple FST in k2.
+
+.. CAUTION::
+
+  There are NO **output labels** in k2. Every arc has a label and you
+  can attach arbitrary attributes with arbitrary name to it.
+
+  If the attached attribute is an N-D tensor, its ``shape[0]`` has to
+  equal the number of arcs in the FSA.
+
+.. NOTE::
+
+  The visualization code handles the attributes ``aux_labels`` specially.
+  Other than this, ``aux_labels`` is like any other attributes attached
+  to the FSA.
+
+Semirings
+---------
+
+k2 supports two kinds of semirings:
```

Comment:
Could edit to 

In the FSA literature, generality is achieved through the concept of "semirings".  The two most common are the "tropical semiring" and "log semiring".  The way we will explain these is a little different from the iterature because we are using the opposite sign.  We won't get into the formalism here, but it relates to what happens when you combine scores from multiple alternative paths.  The two common semirings are:
    - tropical semiring: take the maximum score (or minimum cost)
    - log semiring: log-add the scores (or the negatives of the costs).

Reply:
While k2 only supports these two operations for the core operations, the framework is designed to be flexible through the concept of "attributes" which make it possible to implement the kinds of things that are normally accomplished through exotic semirings such as the Gallic semiring.

Reply:
Thanks, added.